conseil-common/src/main/scala/tech/cryptonomic/conseil/common/rpc/RpcClient.scala

```diff
@@ -0,0 +1,176 @@
+      //                              processing them in any order and with any width of parallelism.
+      // So we also can process the responses in unordered order.
+      .mapAsyncUnordered(maxConcurrent) { chunks =>
+        httpClient
+          .expect[Seq[RpcResponse[R]]](
+            Method.POST(chunks.toList, Uri.unsafeFromString(endpoint), headers: _*)
+          )
+      }
+      .flatMap(Stream.emits)
+      .flatMap {
+        case RpcResponse(_, _, Some(error), _) =>
+          Stream.raiseError[F](error)
+        case RpcResponse(_, Some(result), _, _) =>
+          Stream(result)
+        case response =>
+          Stream.raiseError[F](
+            new UnsupportedOperationException(
+              s"Unexpected response from JSON-RPC server at $endpoint $response"
+            )
+          )
+      }
+
+  /**
+    * Helper method to log debug information.
+    * It wraps scalalogging call into [[Concurrent]] to prevent side effects.
+    * It should be moved to separate logging util in the future.
+    *
+    * @param message Debug message
+    */
+  private def debug(message: String): F[Unit] = Concurrent[F].delay(logger.debug(message))
+}
+
+object RpcClient {
+
+  /**
+    * Create [[cats.Resource]] with [[RpcClient]].
+    *
+    * @param endpoint JSON-RPC server url
+    * @param maxConcurrent Max concurrent processes used to fetch data from JSON-RPC server
+    * @param httpClient Http4s http client
+    * @param headers Additional http headers, e.g. authorization
+    */
+  def resource[F[_]: Concurrent](
+      endpoint: String,
+      maxConcurrent: Int,
+      httpClient: Client[F],
+      headers: Header*
+  ): Resource[F, RpcClient[F]] =
+    for {
+      client <- Resource.pure(
+        new RpcClient[F](
+          endpoint,
+          maxConcurrent,
+          httpClient,
+          headers: _*
+        )
+      )
+    } yield client
+
+  /**
+    * Exception with JSON-RPC error message
+    *
+    * Defined Error Codes:
+    * -32700 ---> parse error. not well formed
+    * -32701 ---> parse error. unsupported encoding
+    * -32702 ---> parse error. invalid character for encoding
+    * -32600 ---> server error. invalid xml-rpc. not conforming to spec.
+    * -32601 ---> server error. requested method not found
+    * -32602 ---> server error. invalid method parameters
+    * -32603 ---> server error. internal xml-rpc error
+    * -32500 ---> application error
+    * -32400 ---> system error
+    * -32300 ---> transport error
+    *
+    * In addition, the range -32099 .. -32000, inclusive is reserved for implementation defined server errors.
+    */
+  case class RpcException(
+      code: Int,
+      message: String,
+      data: Option[String]
+  ) extends RuntimeException(s"Json-RPC error $code: $message ${data.getOrElse("")}")
+      with NoStackTrace
+
+  case class RpcRequest[P](
+      jsonrpc: String,
+      method: String,
+      params: P,
+      id: String
+  )
+
+  case class RpcResponse[R](
```

Comment:
Same story here, shouldn't we limit this `R` to `Json` or maybe `Text` (as `String`) only? Where we are using them?

Reply:
Look at the implicit, the `R` is something that can we encoded and decoded with circe (as part of the http4s), the `R` can be anything as long as you provide the implicits.